Update API for controlling DCC outputs

balazsracz opened the pull request:
Builds an explicit API for controlling DCC outputs of a command station or booster:
- allows more than one output to be present in one application (e.g. main vs program track output, or dual-output boosters)
- makes the railcom cutout configuration have an explicit API
- makes each output have multiple veto bits, each of which can cause the output to be disabled (e.g. short, user via the network, thermal overload, etc). This allows different parts of the application business logic to not need to be aware of each other.
- updates the TivaDCC driver to be able to support 3 independent outputs.
- Removes the POWER_TURNON feature from TivaDCC. It does not work well for certain DCC decoders.

bakerstu commented on src/freertos_drivers/common/DccOutput.hxx:
```diff
@@ -0,0 +1,364 @@
+    };
+
+    /// Disables the output, marking in a bitmask why.
+    virtual void disable_output_for_reason(DisableReason bit) = 0;
```
Add a line break (space)

atanisoft commented:
would it make sense to surface the basic API at a higher level than freertos_drivers/common? Perhaps under dcc or similar? I've implemented a similar structure in ESP32CS recently (splitting OPS and PROG) and can easily shift over to this API.

atanisoft commented on src/freertos_drivers/common/DccOutput.hxx:
```diff
@@ -0,0 +1,364 @@
+    /// @return Bitmask of all currently set disable reasons.
+    virtual uint8_t get_disable_output_reasons() = 0;
+
+    enum class RailcomCutout
```
can you add docs to this enum?

balazsracz replied:
done

atanisoft commented on src/freertos_drivers/common/DccOutput.hxx:
```diff
@@ -0,0 +1,364 @@
+std::atomic_uint8_t DccOutputHw<N>::outputDisableReasons_ {
+    (uint8_t)DccOutput::DisableReason::INITIALIZATION_PENDING};
+template <int N>
+std::atomic_uint8_t DccOutputHw<N>::isRailcomCutoutEnabled_ {2};
```
any reason not to use the enum const here?

balazsracz replied:
Done

balazsracz commented:
> This seems reasonable. Do we have the ability to both have a RailCom cutout on one output while no RailCom cutout and standard bit stream on another?

Yes, there are independent settings of railcom cutout for the different outputs. The DCC driver applies them independently as well. (Of course an application may choose to tie them together for various reasons.)

balazsracz commented:
> would it make sense to surface the basic API at a higher level than freertos_drivers/common? Perhaps under dcc or similar? I've implemented a similar structure in ESP32CS recently (splitting OPS and PROG) and can easily shift over to this API.

That's a good idea. Moved it to dcc/DccOutput.hxx.
